File: 백준/Gold/20207. 달력/달력.py
import sys
input = sys.stdin.readline

N = int(input())
schedule = []
days = [0] * 366

for _ in range(N):
    s, e = map(int, input().split())
    schedule.append((s, e))

schedule.sort(key=lambda x : (x[0], -x[1]))

for day in schedule:
    for i in range(day[0], day[1]+1):
        days[i] += 1

answer = 0

day = 0
pre = -1
max_cnt = 0
while day <= 365:
    if days[day] > 0:
        if pre == -1:
            pre = day
            max_cnt = days[day]
        else:
            max_cnt = max(max_cnt, days[day])
    else:
        if pre != -1:
            answer += (day - pre) * max_cnt
            pre = -1
            max_cnt = 0

    day += 1

if days[365] > 0:
    answer += (day - pre) * max_cnt
print(answer)

File: 백준/Gold/20207. 달력/달력.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.io.IOException;
import java.util.StringTokenizer;
public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st;
        int N = Integer.parseInt(br.readLine());
        int s, e;
        int[] days = new int[366];

        for (int i = 0; i < N; i++) {
            st = new StringTokenizer(br.readLine());
            s = Integer.parseInt(st.nextToken());
            e = Integer.parseInt(st.nextToken());
            for (int j = s; j <= e; j++) {
                days[j]++;
            }
        }

        int answer = 0;
        int day = 0;
        int pre = -1;
        int maxCnt = 0;

        while (day <= 365) {
            if (days[day] > 0) {
                if (pre == -1) {
                    pre = day;
                    maxCnt = days[day];
                } else {
                    maxCnt = Math.max(maxCnt, days[day]);
                }
            } else {
                if (pre != -1) {
                    answer += (day - pre) * maxCnt;
                    pre = -1;
                    maxCnt = 0;
                }
            }

            day++;
        }

        if (days[365] > 0) {
            answer += (day - pre) * maxCnt;
        }

        System.out.println(answer);
    }
}